include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -0,0 +1,252 @@
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    void* p = [&]() {
+      try {
+        return upstream_->allocate(bytes, stream);
+      } catch (std::exception const& e) {
+        RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE {}]",
+                      reinterpret_cast<void*>(stream),
+                      bytes,
+                      e.what());
+        throw;
+      }
+    }();
+
+    // track it.
+    {
+      lock_t lock(mtx);
+      allocations.emplace(p, bytes);
+    }
+
+    return p;
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to by `p`
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes Size of the allocation
+   * @param stream Stream on which to perform the deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    {
+      lock_t lock(mtx);
+      allocations.erase(p);
+    }
+    upstream_->deallocate(p, bytes, stream);
+  }
+
+  /**
+   * @brief Compare the upstream resource to another.
+   *
+   * @throws Nothing.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  bool do_is_equal(device_memory_resource const& other) const noexcept override
+  {
+    if (this == &other)
+      return true;
+    else {
+      tracking_resource_adaptor<Upstream> const* cast =
+        dynamic_cast<tracking_resource_adaptor<Upstream> const*>(&other);
+      if (cast != nullptr)
+        return upstream_->is_equal(*cast->get_upstream());
+      else
+        return upstream_->is_equal(other);
+    }
+  }
+
+  /**
+   * @brief Get free and available memory from upstream resource.
+   *
+   * @throws `rmm::cuda_error` if unable to retrieve memory info.
+   *
+   * @param stream Stream on which to get the mem info.
+   * @return std::pair contaiing free_size and total_size of memory
+   */
+  std::pair<std::size_t, std::size_t> do_get_mem_info(cudaStream_t stream) const override
+  {
+    return upstream_->get_mem_info(stream);
+  }
+
+  struct allocation_info {
+    std::vector<void*> stack_ptrs;
+    std::size_t allocation_size;
+
+    allocation_info() : allocation_size(0){};
+    allocation_info(std::size_t size) : allocation_size(size)
+    {
+      // store off a stack for this allocation
+      const int MaxStackDepth = 64;
+      void* stack[MaxStackDepth];
+      auto depth = backtrace(stack, MaxStackDepth);
+      stack_ptrs.insert(stack_ptrs.end(), &stack[0], &stack[depth]);
```

Comment:
`stack` is a local variable, but you are storing a pointers to it into a vector that outlives it. I think this is a bug, right? Maybe the `&`s should not be here?

Reply:
I'm building an iterator from the array, so we're actually storing the values of the array into the vector. Would you like me to make that more clear with a `std::begin` in there?

Reply:
I misunderstood the code. Why does the array need to be compile-time sized? Why not use std::vector? 

I guess this will change once you use boost::stacktrace.

Reply:
Yep, all different now. It is much cleaner though.